ibtk/include/ibtk/MarkerPatchHierarchy.h

```diff
@@ -146,33 +176,58 @@ class MarkerPatchHierarchy : public SAMRAI::tbox::DescribedClass
      */
     std::size_t getNumberOfMarkers() const;
 
-#if 0
     /**
-     * Save the present state of the object to a SAMRAI database.
+     * Collect all markers on all processors in a single array.
      */
-    void putToDatabase(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> db) override;
-#endif
+    std::pair<EigenAlignedVector<IBTK::Point>, EigenAlignedVector<IBTK::Vector> > collectAllMarkers() const;
 
     /**
      * Set marker velocities with some given velocity field @p u_idx and
      * interaction kernel @p kernel.
      */
     void setVelocities(const int u_idx, const std::string& kernel);
 
+    /**
+     * Advect the markers with their present velocity values by a combination of
+     * the forward and backward Euler methods. For compatibility with various
+     * IBStrategy objects this uses the name `forwardEulerStep` to describe this
+     * timestepping method. After the markers are moved their velocities are
+     * reset to values interpolated from @p u_new_idx via the IB kernel @p
+     * kernel.
+     *
+     * @note This function assumes that @p u_new_idx has up-to-date ghost
+     * values with sufficient width for @p kernel. See LEInteractor for kernel
+     * names.
+     */
+    void forwardEulerStep(const double dt, const int u_new_idx, const std::string& kernel);
+
     /**
      * Advect the markers by some given velocity field at half and new times
      * and the interaction kernel @p kernel.
      *
      * This time integrator is the 'explicit midpoint' method, which uses the
      * present velocity to predict a midpoint position and then uses the
-     * midpoint velocity velocity to compute the new position.
+     * midpoint velocity to compute the new position.
      *
      * @note This function assumes that @p u_half_idx and @p u_new_idx have
      * up-to-date ghost values with sufficient width for @p kernel. See
      * LEInteractor for kernel names.
      */
     void midpointStep(const double dt, const int u_half_idx, const int u_new_idx, const std::string& kernel);
 
+    /**
+     * Advect the markers by some given velocity field at the end of a timestep
+     * and the interaction kernel @p kernel.
+     *
+     * This time integrator is the 'explicit trapezoid' method, which uses the
+     * present velocity to predict a new position and then uses the
+     * trapezoid rule to correct that position.
+     *
+     * @note This function assumes that @p u_new_idx has up-to-date ghost values
+     * with sufficient width for @p kernel. See LEInteractor for kernel names.
+     */
+    void trapezoidalStep(const double dt, const int u_new_idx, const std::string& kernel);
```

Comment:
Don't you need two velocities (u_current and u_new) for trapezoidal step?

Using `J` for the interpolation operator, the explicit trapezoidal rule should be:
```
   X(n+1,~) := X(n) + dt J[X(n)] u(n)
   X(n+1) := 0.5*dt*(J[X(n)] u(n) + J[X(n+1,~)] u(n+1))
```

Reply:
Yes, but like the other one - the present velocity is already known (markers store position and velocity).